### PR TITLE
feat(cli): check user provided genesis if not first run

### LINKED ISSUE #1275: The executable binary still can run normally even the genesis state is changed.
### Contact Details

_No response_

### Current Behavior

This issue is associated with #1274; in #1274, I said the states of the same genesis block are different.

There is another issue: the genesis state is changed, why the executable binary still can run normally?

There are two different scenes:

- Like #1274, the genesis is same but genesis state is changed.

- Users update their `genesis.json` or use an incorrect `genesis.json` to run the chain.

  How to reproduce:
  - Run a dev chain.
  - Edit your `genesis.json` as you like.
  - Run the dev chain again.
  - The chain runs normally even the `genesis` is changed.

### Expected Behavior

The executable binary should be panicked when the genesis state is changed.

### OS

Not Associated

### Axon version

v0.1.0-beta.x

### Kernel

Not Associated

### Relevant log output

_No response_

### Anything else?

It should avoid changing the states, and have checks on that to make sure that any break change could be found before a new commit merged. (ref: I said in #1274)

And also, what I want to say in this issue is, that the executable binaries should do some checks in runtime and throw exceptions when the genesis state was changed.

### PR DESCRIPTION
**Pull request by chaoticlonghair**
**What this PR does / why we need it**:

This PR add a feature that:

- Check user provided genesis if not first run (ref: #1275).

  - Also check if the genesis state is changed.

And this PR fixes a potential issue:

- If it's not the first run, the `genesis` in `Axon` struct is incorrect.

  - Because the `state_root` and `receipts_root` will be set to the correct values only when it is the first run.

    - They are set in the method `execute_transactions(..)`:

      https://github.com/axonweb3/axon/blob/1f346175f773433f69082b03e247e2fb85acd00a/core/run/src/lib.rs#L206-L207

    - But if it's not first run, `execute_transactions(..)` will not be called.
 
      https://github.com/axonweb3/axon/blob/1f346175f773433f69082b03e247e2fb85acd00a/core/run/src/lib.rs#L127-L131
      https://github.com/axonweb3/axon/blob/1f346175f773433f69082b03e247e2fb85acd00a/core/run/src/lib.rs#L149

  This is not a bug currently, because these fields are not used.

**Which issue(s) this PR fixes**:

Fixes #1275.

**Which toolchain this PR adaption**:

No Breaking Change

~**Special notes for your reviewer**~:

~Since it has to run a temporary empty chain to generate the genesis state, so it uses temporary directories to create `RocksAdapter` and `RocksTrieDB`.~

~But [`system_contract::init(..)`] sets a global environment variable, so it has to be called with the real path.~

~I am not sure whether it OK or not:~
- ~Could [`system_contract::init(..)`] be called twice?~
- ~Is [`system_contract::init(..)`] sensitive with the height?~
  ~Because the first time it is called, the height will always be `0` after this PR.~

**CI Description**

| CI Name                                   | Description                                                     |
| ----------------------------------------- | --------------------------------------------------------------- |
| *Chaos CI*                                | Test the liveness and robustness of Axon under terrible network condition    |
| *Cargo Clippy*                            | Run `cargo clippy --all --all-targets --all-features`      |
| *Coverage Test*                           | Get the unit test coverage report                             |
| *E2E Test*                                | Run end-to-end test to check interfaces                         |
| *Code Format*                             | Run `cargo +nightly fmt --all -- --check` and `cargo sort -gwc`     |
| *Web3 Compatible Test*                    | Test the Web3 compatibility of Axon                               |
| *v3 Core Test*                            | Run the compatibility tests provided by Uniswap V3             |
| *OCT 1-5 \| 6-10 \| 11 \| 12-15 \| 16-19* | Run the compatibility tests provided by OpenZeppelin           |

**CI Usage**

> Check the CI you want to run below, and then comment `/run-ci`.

**CI Switch**

- [ ] Chaos CI
- [x] Cargo Clippy
- [ ] Coverage Test
- [x] E2E Tests
- [x] Code Format
- [x] Unit Tests
- [ ] Web3 Compatible Tests
- [ ] OCT 1-5 And 12-15
- [ ] OCT 6-10
- [ ] OCT 11
- [ ] OCT 16-19
- [ ] v3 Core Tests

[`system_contract::init(..)`]: https://github.com/axonweb3/axon/blob/1f346175f773433f69082b03e247e2fb85acd00a/core/executor/src/system_contract/mod.rs#L80

### CODE REVIEW

**Comment by Flouse** [comment]
/run-ci

=> [CI Error: Must have push access to view collaborator permission.](https://github.com/axonweb3/axon/actions/runs/5599551094/jobs/10240662319#step:2:11)

File: core/run/src/lib.rs [hunk 1]
```diff
@@ -979,6 +998,72 @@ where
...
+                "The {}th tx in genesis execute failed, reason {:?}",
+                i, r.exit_reason
+            );
+        }
+    });
+
+    Ok(resp)
+}
+
+async fn save_block(
```
Comment (KaoImin) [inline core/run/src/lib.rs]: Missing save receipts here.

Comment (chaoticlonghair) [reply core/run/src/lib.rs]: Copy that! I will update after collecting more review suggestions.

p.s. I just copied the old code, so this bug is existed in current main branch.

Comment (KaoImin) [reply core/run/src/lib.rs]: Yes, it is existed in current main. We will fix it in this migration.

Comment (chaoticlonghair) [reply core/run/src/lib.rs]: I will fix this in Next Next PR.

Because the `migrate` sub-command should be merged before the bugfix.

**Comment by chaoticlonghair** [comment]
ping @axonweb3/axon-dev

**Review by driftluo — APPROVED** [review]

**Review by KaoImin — APPROVED** [review]

**Pull request merged**